File: rust/src/data_structures/red_black_tree.rs
use std::{ops::Not, ptr::NonNull};

#[derive(Eq, PartialEq, Clone, Copy)]
enum Side {
    Left,
    Right,
}

#[derive(Eq, PartialEq)]
enum Color {
    Red,
    Black,
}

struct TreeNode<T> {
    value: T,
    color: Color,
    left: Option<NonNull<TreeNode<T>>>,
    right: Option<NonNull<TreeNode<T>>>,
    parent: Option<NonNull<TreeNode<T>>>,
    is_root: bool
}

#[derive(Eq, PartialEq, Ord, PartialOrd)]
pub struct RedBlackTree<T> {
    root: Option<NonNull<TreeNode<T>>>,
    length: usize
}

impl Not for Side {
    type Output = Side;

    fn not(self) -> Self::Output {
        match self {
            Side::Left => Side::Right,
            Side::Right => Side::Left,
        }
    }
}

impl<T> TreeNode<T> {
    fn new(value: T) -> Self {
        TreeNode {
            value,
            color: Color::Red,
            left: None,
            right: None,
            parent: None,
            is_root: false
        }
    }

    fn root(value: T) -> Self {
        TreeNode {
            value,
            color: Color::Black,
            left: None,
            right: None,
            parent: None,
            is_root: true
        }
    }

    fn change_color(&mut self) {
        self.color = match self.color {
            Color::Black => Color::Red,
            Color::Red => Color::Black
        };
    }

    fn child(&mut self, side: Side) -> &Option<NonNull<TreeNode<T>>> {
        match side {
            Side::Left => &self.left,
            Side::Right => &self.right,
        }
    }

    fn child_mut(&mut self, side: Side) -> &mut Option<NonNull<TreeNode<T>>> {
        match side {
            Side::Left => &mut self.left,
            Side::Right => &mut self.right,
        }
    }

    fn check_color_match(&self) -> Option<Side> {
        let left_match = self
            .left
            .as_ref()
            .map_or(false, |&left_non_null| unsafe {
                self.color == Color::Red && self.color == left_non_null.as_ref().color
            });

        if left_match {
            return Some(Side::Left);
        }

        let right_match = self
            .right
            .as_ref()
            .map_or(false, |&right_non_null| unsafe {
                self.color == Color::Red && self.color == right_non_null.as_ref().color
            });

        if right_match {
            return Some(Side::Right);
        }

        None // no rotations are needed
    }
}

impl<T: Ord + PartialOrd> RedBlackTree<T> {
    pub fn new() -> Self {
        RedBlackTree {
            root: None,
            length: 0
        }
    }

    pub fn insert(&mut self, value: T) -> bool {
        let inserted;
        if let Some(mut root) = self.root {
            unsafe {
                inserted = insert_value(&mut root, value);
                self.update_root();
                self.length += 1;
            }
        } else {
            let root_box = Box::new(TreeNode::<T>::root(value));
            let root_ptr = NonNull::new(Box::into_raw(root_box));
            self.root = root_ptr;
            self.length += 1;
            inserted = true;
        }
        inserted
    }

    fn update_root(&mut self) {
        unsafe {
            while let Some(mut root) = self.root {
                if let Some(mut parent) = root.as_ref().parent {
                    root.as_mut().is_root = false;
                    parent.as_mut().is_root = true;
                    self.root = Some(parent);
                } else {
                    break;
                }
            }
        }
    }
}


unsafe fn insert_value<T: Ord + PartialOrd>(node: &mut NonNull<TreeNode<T>>, value: T) -> bool {
    let node_ptr = node.as_mut();
    if node_ptr.value == value {
        return false; // Value already exists in the tree
    }

    let target_node = 
        if value < node_ptr.value { 
            &mut node_ptr.left 
        } else { 
            &mut node_ptr.right 
        };

    let inserted = match target_node {
        Some(mut ptr) => insert_value(&mut ptr, value),
        None => {
            let mut new_node = TreeNode::<T>::new(value);
            new_node.parent = Some(*node);

            let node_box = Box::new(new_node);
            let node_ptr = NonNull::new(Box::into_raw(node_box));
            *target_node = node_ptr;
            true
        }
    };

    let match_side = node.as_ref().check_color_match(); // Return 'Side' if node and new_node are with same 'Red' color
    if match_side.is_some() {
        // TODO:
    }
    
    inserted
}

File: csharp/RedBlackTree/RedBlackTree.cs
namespace RedBlackTree;

public enum Side
{
    Left, Right,
}

public enum Color
{
    Red, Black,
}

public static class SideExtensions
{
    public static Side Switch(this Side side)
        => side switch
        {
            Side.Left => Side.Right,
            Side.Right => Side.Left,
            _ => throw new ArgumentOutOfRangeException(nameof(side), side, null),
        };
}

public class TreeNode<T> where T : IComparable<T>
{
    public T Value;
    public Color Color;
    public TreeNode<T> Left;
    public TreeNode<T> Right;
    public TreeNode<T> Parent;
    public bool IsRoot;

    internal TreeNode(T value, TreeNode<T> parent = null, bool isRoot = false)
    {
        Value = value;
        Parent = parent;
        if (isRoot)
        {
            IsRoot = true;
            Color = Color.Black;
        }
    }

    internal TreeNode<T> GetChildNode(Side side)
    {
        var node = side switch
        {
            Side.Left => this.Left,
            Side.Right => this.Right,
            _ => throw new ArgumentException(null, nameof(side)),
        };

        return node;
    }

    internal void ChangeColor()
    {
        this.Color = this.Color switch
        {
            Color.Red => Color.Black,
            Color.Black => Color.Red,
            _ => throw new ArgumentException(null, nameof(Color)),
        };
    }

    internal Side? CheckColorMatch()
    {
        if (Left != null && Color == Color.Red && Color == Left.Color)
            return Side.Left;

        if (Right != null && Color == Color.Red && Color == Right.Color)
            return Side.Right;

        return null;
    }
}

public class RedBlackTree<T> where T : IComparable<T>
{
    public TreeNode<T> Root;
    public uint Length { get; private set; }

    public RedBlackTree()
    {
        Root = null;
        Length = 0;
    }

    public bool Insert(T value)
    {
        bool inserted = false;
        if (Root != null)
        {
            inserted = InsertNode(Root, value, inserted);
        }
        else
        {
            Root = new TreeNode<T>(value, isRoot: true);
            inserted = true;
        }

        if (inserted)
        {
            Length++;
        }
        return inserted;
    }

    private bool InsertNode(TreeNode<T> node, T value, bool inserted)
    {
        TreeNode<T> insertedNode = null;
        if (node.Value.CompareTo(value) == 0)
        {
            inserted = false;
            return inserted;
        }

        if (value.CompareTo(node.Value) < 0)
        {
            if (node.Left != null)
                inserted = InsertNode(node.Left, value, inserted);
            else
            {
                inserted = true;
                node.Left = new TreeNode<T>(value, parent: node);
                insertedNode = node.Left;
            }
        }
        else
        {
            if (node.Right != null)
                inserted = InsertNode(node.Right, value, inserted);
            else
            {
                inserted = true;
                node.Right = new TreeNode<T>(value, parent: node);
                insertedNode = node.Right;
            }
        }

        if (inserted)
        {
            var colorMatch = node.CheckColorMatch();
            if (colorMatch.HasValue)
            {
                CheckForRebalancing(insertedNode, colorMatch.Value);
            }
        }

        return inserted;
    }

    private void CheckForRebalancing(TreeNode<T> node, Side side)
    {
        if (node.Parent == null) return;

        if (node.Color == Color.Red && node.Parent.Color == Color.Red)
        {
            TreeNode<T> grandParent = node.Parent?.Parent;
            TreeNode<T> uncle = null;

            if (grandParent != null)
            {
                if (node.Parent == grandParent.Left)
                    uncle = grandParent.Right;
                else
                    uncle = grandParent.Left;
            }

            if (uncle != null && uncle.Color == Color.Red)
            {
                node.Parent.Color = Color.Black;
                uncle.Color = Color.Black;
                if (grandParent != null)
                {
                    grandParent.Color = grandParent.IsRoot ? Color.Black : Color.Red;
                    CheckForRebalancing(grandParent, side);
                }
            }
            else
            {
                if (uncle == null || uncle.Color == Color.Black)
                {
                    if (grandParent != null)
                    {
                        if (node.Parent != grandParent.GetChildNode(side))
                        {
                            Rotate(side, node.Parent);
                            side = side.Switch();
                        }

                        Rotate(side, grandParent);
                        node.ChangeColor();
                        node.GetChildNode(side.Switch()).ChangeColor();
                    }
                }

                Root.Color = Color.Black; // Ensure root is always black
            }
        }
    }

    private void Rotate(Side side, TreeNode<T> node)
    {
        // Save child and grandchild nodes
        TreeNode<T> child = node.GetChildNode(side);
        TreeNode<T> grandChild = child?.GetChildNode(side.Switch());

        // Perform rotation
        if (node.IsRoot)
        {
            Root = child;
            child.Parent = null;
            child.IsRoot = true;
        }
        else
        {
            TreeNode<T> parent = node.Parent;
            if (parent.Left == node)
                parent.Left = child;
            else
                parent.Right = child;

            child.Parent = parent;
        }

        node.Parent = child;
        if (side == Side.Left)
        {
            child.Right = node;
            node.Left = grandChild;
        }
        else
        {
            child.Left = node;
            node.Right = grandChild;
        }

        // Reassign grandchild
        if (grandChild != null)
            grandChild.Parent = node;
    }
}
